### PR TITLE
shared string repo -- WIP

### PR DESCRIPTION
**Pull request by havardpe**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@arnej27959 please review

### CODE REVIEW

**Review by arnej27959 — APPROVED** [review]
looks very promising

File: vespalib/src/vespa/vespalib/util/shared_string_repo.h [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            uint32_t local_hash = full_hash >> PART_BITS;
+            uint32_t local_idx = _partitions[part].resolve(AltKey{str, local_hash});
+            return (((local_idx << PART_BITS) | part) + 1);
+        } else {
+            return 0;
+        }
+    }
+
+    static vespalib::string get(uint32_t id) {
+        uint32_t part = (id - 1) & PART_MASK;
```
Comment (arnej27959) [inline vespalib/src/vespa/vespalib/util/shared_string_repo.h:145]: don't you need to check for `id == 0` here?

Comment (havardpe) [reply vespalib/src/vespa/vespalib/util/shared_string_repo.h:145]: yes, I do

**Review by arnej27959 — APPROVED** [review]
👍

**Review by arnej27959 — APPROVED** [review]
LGTM

**Pull request merged**